src/core_read.cpp

```diff
@@ -108,38 +108,44 @@ bool CheckTxScriptsSanity(const CMutableTransaction& tx)
     return true;
 }
 
-bool DecodeHexTx(CMutableTransaction& tx, const std::string& strHexTx, bool fTryNoWitness)
+bool DecodeHexTx(CMutableTransaction& tx, const std::string& hex_tx, bool try_no_witness, bool try_witness)
 {
-    if (!IsHex(strHexTx)) {
+    if (!IsHex(hex_tx)) {
         return false;
     }
 
-    std::vector<unsigned char> txData(ParseHex(strHexTx));
+    std::vector<unsigned char> txData(ParseHex(hex_tx));
 
-    if (fTryNoWitness) {
+    if (try_no_witness) {
         CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS);
         try {
             ssData >> tx;
-            if (ssData.eof() && CheckTxScriptsSanity(tx)) {
+            if (ssData.eof() && (CheckTxScriptsSanity(tx) || !try_witness)) {
                 return true;
             }
         }
         catch (const std::exception&) {
+            if(!try_witness) {
+                return false;
+            }
+
             // Fall through.
         }
     }
 
-    CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
-    try {
-        ssData >> tx;
-        if (!ssData.empty()) {
+    if (try_witness) {
+        CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
+        try {
+            ssData >> tx;
+            if (!ssData.empty()) {
+                return false;
+            }
+        }
+        catch (const std::exception&) {
             return false;
         }
     }
-    catch (const std::exception&) {
-        return false;
-    }
-
+    
     return true;
```

Comment:
If you move this `return true;` inside the `if (try_witness)` block, the `if(!try_witness)` inside the `catch` above becomes unnecessary. I think that'd be a cleaner flow.

Reply:
Good idea yep, done